pkg/networkserver/utils.go

```diff
@@ -378,6 +379,32 @@ func publishEvents(ctx context.Context, evs ...events.Event) {
 	events.Publish(evs...)
 }
 
+func uplinkDwellTime(dev *ttnpb.EndDevice, fp *frequencyplans.FrequencyPlan, phy *band.Band) bool {
+	switch {
+	case dev.GetMacState().GetCurrentParameters().GetUplinkDwellTime() != nil:
+		return dev.MacState.CurrentParameters.UplinkDwellTime.Value
+	case fp.DwellTime.Uplinks != nil:
+		return *fp.DwellTime.Uplinks
+	case phy.BootDwellTime.Uplinks != nil:
+		return *phy.BootDwellTime.Uplinks
+	default:
+		return false
+	}
+}
+
+func downlinkDwellTime(dev *ttnpb.EndDevice, fp *frequencyplans.FrequencyPlan, phy *band.Band) bool {
+	switch {
+	case dev.GetMacState().GetCurrentParameters().GetDownlinkDwellTime() != nil:
+		return dev.MacState.CurrentParameters.DownlinkDwellTime.Value
+	case fp.DwellTime.Downlinks != nil:
+		return *fp.DwellTime.Downlinks
+	case phy.BootDwellTime.Downlinks != nil:
+		return *phy.BootDwellTime.Downlinks
+	default:
+		return false
+	}
+}
```

Comment:
That leads me here. In those Asian countries that use the generic AS frequency plans that do not configure dwell time, and where we do not know a boot value because (RP1), we assume that dwell time is deactivated here. That may very well lead to downlinks that are too long.

Or am I missing anything?

Reply:
You are correct, and not missing anything. 

We will not control the dwell time setting here, because frankly we don't know what value to set this to. I think that the frequency plan should include dwell time settings, explicitly, if it really wants a particular state - enabled or disabled - to be achieved. If we force the desired parameters to always include a conservative value (uplink disabled, downlink enabled), we're also affecting other bands. 

I propose the following:
- The frequency plans should explicitly enable or disable dwell time limitations
- If end devices have restrictive downlink dwell time limitations enabled at boot time, they should explicitly set these boot time settings via the MAC settings
- We assume the worst case scenario here and instead of defaulting to false, optimistically, we default to true, pessimistically.

Reply:
> I propose the following:
> 
> * The frequency plans should explicitly enable or disable dwell time limitations

I agree, but in generic plans we can't define dwell time. Those plans are used by RP2 versions too, that _do_ set a default. So if we set a default in the frequency plan, it should only be for RP1, and so that would mean we would need to prioritize the PHY defaults over the FP, which we shouldn't do.

So what we are looking for is a sane default if the end device uses a generic AS band plan (with dwell time unset), the device uses RP1 (with dwell time unset) and the dwell time has not been configured through MAC settings. This criteria match nearly 100% of AS923 devices today (not including AS923 in Australia).

This was the reason to go with a default that is widely deployed in the field, namely the one used by LoRaMac-node. That library or derivatives thereof are effectively used in most end devices, or are used as reference for implementations.

Other examples:
- Arm MbedOS: https://github.com/ARMmbed/mbed-os/blob/master/connectivity/lorawan/lorastack/phy/LoRaPHYAS923.cpp#L287-L288
- Arduino LMiC: https://github.com/mcci-catena/arduino-lmic/blob/master/src/lmic/lorabase_as923.h#L99-L102

@htdvisser do you still think we shouldn't go with these boot parameters in RP1?

> * If end devices have restrictive downlink dwell time limitations enabled at boot time, they should explicitly set these boot time settings via the MAC settings

All AS923 devices based on LoRaMac-node, MbedOS or MCCI's LMiC will have dwell time enabled at boot time, and we are introducing a new MAC setting that should be enabled for end devices. So you are suggesting to explicitly enable downlink dwell time for all these devices.

We don't have support for this in the Device Repository and not in the Console, let alone as default value with tooltip etc. The CLI flag wouldn't also be enabled by default. In the Device Repository we would need to verify with device makers that their AS923 devices use LoRaMac-node, MbedOS or LMiC and if so, enable dwell time in the MAC settings. Still, we would need to apply that to all existing devices created.

> * We assume the worst case scenario here and instead of defaulting to false, optimistically, we default to true, pessimistically. I don't feel this is very efficient, but it is what it is.

Yes, so if we don't go with a sane boot parameter for RP1, we should return true here as default and use the most restrictive setting. We should also _always_ send `TxParamSetupReq` to make sure the end device then gets dwell time enabled. If we don't do that, and the end device does not use downlink dwell time and we do (because of `true`), downlink may be broken because the RX1 DR offset isn't correct (RX1 response to DR0/1 with DR2 I believe).

Reply:
So we have two options when using the generic AS923 plans:

1. Use dwell time boot parameters in RP1 based on the default settings by all major LoRaWAN end device libraries. End devices that do not use these defaults should use the new MAC settings (to override the current params) or a frequency plan that explicitly sets the dwell time (to override the desired params)
2. Always enable dwell time and always configure this explicitly with `TxParamSetupReq`. This is inefficient.

Reply:
> This was the reason to go with a default that is widely deployed in the field, namely the one used by LoRaMac-node. That library or derivatives thereof are effectively used in most end devices, or are used as reference for implementations.
> [...]
> @htdvisser do you still think we shouldn't go with these boot parameters in RP1?

With V2 we've made the mistake of following the defaults of certain implementations instead of the specification. Let's not make the same mistake now.

I don't know the details of the implementation in the NS, but perhaps it's possible to just accept that we don't know, so accepting that `UplinkDwellTime` and `DownlinkDwellTime` can be `1`, `0` or `unknown` until we actually need to know.

It's clear what we need to do for `1` and `0`. For `unknown`, if we only receive uplinks <400ms, we don't need to do anything. If we receive an uplink >400ms, we know that the device thinks that `UplinkDwellTime=0`, and we can adjust if needed. If we only need to send downlinks <400ms, we don't need to do anything. If we need to send a downlink >400ms to an RP1 device (nb: RP2 devices accept >400ms downlinks even if `DownlinkDwellTime=1`) we first need to adjust the params so that we're sure the device will accept the downlink.

Reply:
> With V2 we've made the mistake of following the defaults of certain implementations instead of the specification. Let's not make the same mistake now.

There is no "instead of the specification" here, because there is nothing in the specification. In RP1 this was left open, and we see that the three major LoRaWAN end device stacks use a different implementation than The Things Stack.

A thing that we learned from V2 is that we should not guess things based on perceived behavior. Uplink and downlink dwell time are not necessarily the same. For downlink dwell time, even if we have less than 400 ms to send, we have to deal with the RX1 data rate offset. NS doesn't know which DR to use when the end device used DR0 or DR1 for uplink:

<img width="919" alt="Screenshot 2022-01-27 at 18 08 29" src="https://user-images.githubusercontent.com/13334001/151408588-87467a2a-f36b-4b80-a545-9d94b0acfb23.png">

For uplink, guessing whether uplink dwell time is enabled or not based on uplink message length isn't ideal. But apart from regional compliance checking, this isn't too important.

For downlink, we really need to send `TxParamSetupReq` and wait for `TxParamSetupAns` before we have a reliable downlink path.

---

I'd propose the following combination:

1. Reinstate the defaults based on boot parameters. This makes RX1 more reliable when the device uses DR0 or DR1. We need this RX1 for `TxParamSetupReq`. This configures current parameters. The value in the device's MAC settings takes priority. I still think this is a better default than what we have now, and there's no way around choosing a setting before we got `TxParamSetupAns`
2. When the frequency plan does not have a value for dwell time set, assume it is `true` for both directions: the most restrictive case. This configures the desired parameters
3. Send `TxParamSetupReq` after all MAC state resets, unless the dwell time settings of the device's MAC settings (so _not_ the current parameters) match the desired parameters already
4. Add frequency plans for HK and Taiwan with dwell time activated

Reply:
1. :heavy_check_mark: re-added and tests fixed
2. Since we're going to default to `true` in the desired parameters, this will enable bidirectional dwell time explicitly in AU915 for 1.0.3+ devices. Do we want this ?
For reference, past behavior so far was to assume false bidirectionally.
3. The only difference between the MAC settings and the current parameters is the boot defaults. So we would like to send `TxParamSetupReq` even though the boot defaults match the desired parameters. 
Am I correct to assume that we're doing this in order to ensure that the dwell time limitations are set correctly for devices that do not use our assumed defaults ?
4. I assume this is for the Gateway Server mainly, since the NS will use `true` anyway due to point 2. I can follow up on this.

Reply:
> 2\. Since we're going to default to `true` in the desired parameters, instead of `nil`, this will enable bidirectional dwell time explicitly in AU915 for 1.0.3+ devices. Do we want this ?
>     For reference, past behavior so far was to assume false bidirectionally.

Isn't that an issue in the frequency plan then?

> 4\. I assume this is for the Gateway Server mainly, since the NS will use `true` anyway due to point 2. I can follow up on this.

Yeah indeed, this is less important. It's good to be explicit. No need to address this now.

> Extra question: what value do we assume for devices for which the current parameters are `nil`, in case there are no boot defaults ? `true` or `false` ?
> [...]
> so maybe falling back to the frequency plan if there are no boot time settings is not evil here.

Yes I think we should that in this case.

But aren't there always boot defaults?

Reply:
> Isn't that an issue in the frequency plan then?

The RP document indeed marks this as a 'regional dependence', so indeed we will fix this using new frequency plans.

> But aren't there always boot defaults?

I've added boot defaults only to bands that explicitly have the text `<band> end-devices SHALL always consider *DwellTime = <value>`, and US915. 

For some AS923 combinations for which there never was a release of LoRaMac-node, we don't really have boot defaults, nor any frequency plan dwell time settings (in the generic plan). We will enable dwell time now for these versions, due to the desired `true` fallback, but we have to fit into RX2 worst case (due to the possible offset mismatch).